File: sdk/src/Collection.ts
import {merge} from 'lodash'
import {rpc, wallet} from '@cityofzion/neon-core'
import {sc} from "@cityofzion/neon-js";
import {CollectionType, NetworkOption, PropConstructorOptions} from "./interface";
import {CollectionAPI} from "./api";
import fs from 'fs'
import {ContractParamLike} from "@cityofzion/neon-core/lib/sc";

const DEFAULT_OPTIONS: PropConstructorOptions = {
  network: NetworkOption.LocalNet
}

/**
 * The Collection prop is designed to store static-immutable data for reference in other projects. Storing static data
 * in contracts is very expensive and inefficient, especially for new projects.  This contract resolves that issue by creating
 * library for static data. This class exposes the interface along with a number of helpful features to make the smart
 * contract easy to use for typescript developers.
 *
 * All of the prop helper classes will auto-configure your network settings.  The default configuration will interface with
 * the contract compiled with this project and deployed locally at http://localhost:50012.  For more information on deploying
 * contract packages, refer to the quickstart.
 *
 * All methods support a signer.  If the method can be run as a test-invoke, optionally populating the signer parameter
 * will publish the invocation and return the txid instead of the method response.
 *
 * To use this class:
 * ```typescript
 * import {Collection} from "../../dist" //import {Collection} from "@cityofzion/props
 *
 * const collection: Collection = new Collection()
 * await collection.init() // interfaces with the node to resolve network magic
 *
 * const total = await collection.totalCollections()
 * console.log(total) // outputs the total collection count in the contract
 * ```
 */
export class Collection {
  private options: PropConstructorOptions = DEFAULT_OPTIONS
  private networkMagic: number = -1

  constructor(options: PropConstructorOptions = {}) {
    switch(options.network) {
      case NetworkOption.TestNet:
        this.options.node = 'https://testnet1.neo.coz.io:443'
        this.options.scriptHash = '0xe246d4c03d4d8c3fb529add124289e81a8eb6b6a'
        break
      case NetworkOption.MainNet:
        this.options.node = 'https://mainnet1.neo.coz.io:443'
        this.options.scriptHash = '0xf05651bc505fd5c7d36593f6e8409932342f9085'
        break
      default:
        this.options.node = 'http://127.0.0.1:50012'
        this.options.scriptHash = '0xffa8d18f1feec460d301bd755fabab793be4b7cb'
        break
    }
    this.options = merge({}, this.options, options)
  }

  /**
   * Gets the magic number for the network and configures the class instance.
   */
  async init() {
    const getVersionRes = await this.node.getVersion()
    this.networkMagic = getVersionRes.protocol.network
  }

  /**
   * The the node that the instance is connected to.
   */
  get node(): rpc.RPCClient {
    if (this.options.node) {
      return new rpc.RPCClient(this.options.node)
    }
    throw new Error('no node selected!')
  }

  /**
   * The contract script hash that is being interfaced with.
   */
  get scriptHash() : string {
    if (this.options.scriptHash) {
      return this.options.scriptHash
    }
    throw new Error('scripthash defined')
  }

  /**
   * Publishes an array of immutable data to the smart contract along with some useful metadata.
   *
   * @param description A useful description of the collection.
   * @param collectionType The type of the data being store.  This is an unregulated field.  Standard NVM datatypes should
   * adhere to existing naming conventions.
   * @param extra An unregulated field for unplanned feature development.
   * @param values An array of values that represent the body of the collection.
   * @param signer The signer of the transaction.
   *
   * @returns A transaction ID.  Refer to {@link helpers.txDidComplete} for parsing.
   */
  async createCollection(description: string, collectionType: string, extra: string, values: string[], signer: wallet.Account): Promise<string> {
    return CollectionAPI.createCollection(this.node.url, this.networkMagic, this.scriptHash, description, collectionType, extra,  values, signer)
  }

  /**
   * Loads a {@link CollectionType} formatted JSON file and pushes it to the smart contract.
   *
   * @param path The path to the file.
   * @param signer The signer of the transaction.
   *
   * @returns A transaction ID. Refer to {@link helper.txDidComplete} for parsing.
   */
  async createFromFile(path: string, signer: wallet.Account): Promise<string> {
    const localCollection = JSON.parse(fs.readFileSync(path).toString()) as CollectionType
    const formattedValues = (localCollection.values as any[]).map( (value: string | number) => {
      switch (localCollection.type) {
        case 'string':
          return sc.ContractParam.string(value as string)
        case 'int':
          return sc.ContractParam.integer(value as number)
      }
    }) as ContractParamLike[]
    return CollectionAPI.createCollectionRaw(this.node.url, this.networkMagic, this.scriptHash, localCollection.description, localCollection.type, localCollection.extra, formattedValues, signer)
  }

  /**
   * Gets a JSON formatting collection from the smart contract.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The requested collection **OR** a txid if the signer parameter is populated.
   */
  async getCollectionJSON(collectionId: number, signer?: wallet.Account): Promise<CollectionType | string> {
    return CollectionAPI.getCollectionJSON(this.node.url, this.networkMagic, this.scriptHash, collectionId, signer)
  }

  /**
   * Gets the bytestring representation of the collection.  This is primarilly used for inter-contract interfacing,
   * but we include it here for completeness.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The bytestring representation of the collection. **OR** a txid if the signer parameter is populated.
   */
  async getCollection(collectionId: number, signer?: wallet.Account): Promise<string> {
    return CollectionAPI.getCollection(this.node.url, this.networkMagic, this.scriptHash, collectionId, signer)
  }

  /**
   * Returns the value of a collection from a requested index.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param index The index of the array element being requested.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The value of the collection element **OR** a txid if the signer parameter is populated.
   */
  async getCollectionElement(collectionId: number, index: number, signer?: wallet.Account): Promise<string> {
    return CollectionAPI.getCollectionElement(this.node.url, this.networkMagic, this.scriptHash, collectionId, index, signer)
  }

  /**
   * Gets the array length of a requested collection.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The length of the collection **OR** a txid if the signer parameter is populated.
   */
  async getCollectionLength(collectionId: number, signer?: wallet.Account): Promise<number> {
    return CollectionAPI.getCollectionLength(this.node.url, this.networkMagic, this.scriptHash, collectionId, signer)
  }

  /**
   * Gets the values of a collection, omitting the metadata.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The values in the collection **OR** a txid if the signer parameter is populated.
   */
  async getCollectionValues(collectionId: number, signer?: wallet.Account): Promise<string[] | any> {
    return CollectionAPI.getCollectionValues(this.node.url, this.networkMagic, this.scriptHash, collectionId, signer)
  }

  /**
   * Maps byte entropy onto a collection's values and returns the index of the result.  The mapping is made as follows:
   *
   * [0 -> MAX(entropyBytes.length)][entropy] -> [0 -> collection.length][index]
   *
   * This method is primarily useful for computationally efficient contract interfacing. For random sampling, or
   * sampling from a distribution, use {@link getCollectionLength} in combination with {@link getCollectionElement} or
   * {@link sampleFromCollection}.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param entropy Bytes to use for the mapping.
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The element from the mapping **OR** a txid if the signer parameter is populated.
   */
  async mapBytesOntoCollection(collectionId: number, entropy: string, signer?: wallet.Account): Promise<string> {
    return CollectionAPI.mapBytesOntoCollection(this.node.url, this.networkMagic, this.scriptHash, collectionId, entropy, signer)
  }

  /**
   * Samples a uniform random value from the collection using a Contract.Call to the {@link Dice} contract.
   *
   * @param collectionId The collectionID being requested.  Refer to {@link https://props.coz.io} for a formatted list.
   * @param samples The number of samples to return
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns A uniform random sample from the collection. **OR** a txid if the signer parameter is populated.
   * **Note:** This method will not randomly generate unless the transaction is published so use the signer field for
   * testing.
   */
  async sampleFromCollection(collectionId: number, samples: number, signer?: wallet.Account): Promise<string> {
    return CollectionAPI.sampleFromCollection(this.node.url, this.networkMagic, this.scriptHash, collectionId, samples, signer)
  }

  /**
   * Samples uniformly from a collection provided at the time of invocation.  Users have the option to 'pick', which
   * prevents a value from being selected multiple times.  The results are published as outputs on the transaction.
   * @param values an array of values to sample from
   * @param samples the number of samples to fairly select from the values
   * @param pick Are selected values removed from the list of options for future samples?
   * @param signer The signer of the transaction.
   */
  async sampleFromRuntimeCollection(values: string[], samples: number, pick: boolean, signer: wallet.Account): Promise<string> {
    return CollectionAPI.sampleFromRuntimeCollection(this.node.url, this.networkMagic, this.scriptHash, values, samples, pick, signer)
  }

  /**
   * Gets the total collections.  Collection IDs are autogenerated on range [1 -> totalCollections] inclusive if you are
   * planning to iterate of their collection IDs.
   *
   * @param signer An optional signer. Populating this field will publish the transaction and return a txid instead of
   * running the invocation as a test invoke.
   *
   * @returns The total number of collections stored in the contract. **OR** a txid if the signer parameter is populated.
   */
  async totalCollections(signer?: wallet.Account): Promise<number | undefined> {
    return CollectionAPI.totalCollections(this.node.url, this.networkMagic, this.scriptHash, signer)
  }

  async update(script: string, manifest: string, signer: wallet.Account): Promise<string | undefined> {
    return CollectionAPI.update(this.node.url, this.networkMagic, this.scriptHash, script, manifest, '', signer)
  }

}

File: contracts/collection/collection.py
from typing import Any, Dict, cast
from boa3.builtin.type import UInt160
from boa3.builtin import CreateNewEvent, NeoMetadata, metadata, public, contract
from boa3.builtin.interop.blockchain import Transaction
from boa3.builtin.interop.runtime import script_container
from boa3.builtin.interop.contract import update_contract
from boa3.builtin.interop.storage import get, put
from collection.helpers.collection import Collection, create_collection_internal, get_collection_json_internal, \
    get_collection_internal, total_collections_internal


OWNER_KEY = b'OWNER'


@metadata
def manifest_metadata() -> NeoMetadata:
    """
    Defines this smart contract's metadata information
    """
    meta = NeoMetadata()
    meta.author = "COZ, Inc."
    meta.description = "A public smart contract for storing immutable collections of stuff"
    meta.email = "contact@coz.io"
    meta.supported_standards = []
    return meta


new_collection = CreateNewEvent(
    [
        ('collection_id', bytes),
    ],
    'NewCollection'
)


@public
def create_collection(description: bytes, collection_type: bytes, extra: bytes,  vals: [bytes]) -> int:
    """
    Creates a new collection
    :param description: A brief string formatted description of the collection
    :param collection_type: A string formatted type definition.  Where possible, use naming the Neo type naming syntax
    :param extra: An option payload for supplemental data
    :param vals: The array of values in the collection
    :return: The collection id
    """
    tx = cast(Transaction, script_container)
    author: UInt160 = tx.sender

    collection_id: bytes = create_collection_internal(author, description, collection_type, extra, vals)
    new_collection(collection_id)
    return collection_id.to_int()


@public
def get_collection_json(collection_id: bytes) -> Dict[str, Any]:
    """
    Gets the JSON formatted collection
    :param collection_id: The byte formatted collection_id
    :return: A dictionary representing the collection
    """
    return get_collection_json_internal(collection_id)


@public
def get_collection(collection_id: bytes) -> Collection:
    """
    Gets a Collection class instance
    :param collection_id: The byte formatted collection_id
    :return: A Collection instance of the requested collection_id
    """
    return get_collection_internal(collection_id)


@public
def get_collection_element(collection_id: bytes, index: int) -> bytes:
    """
    Gets the value at a requested index in a collection
    :param collection_id: The collection_id to reference
    :param index: The index of the value to get
    :return: bytes at the requested index
    """
    collection: Collection = get_collection_internal(collection_id)
    return collection.get_value(index)


@public
def get_collection_length(collection_id: bytes) -> int:
    """
    Gets the length of a collection
    :param collection_id: The collection_id to get the length of
    :return: An integer representing the length of the collection
    """
    collection_values: [bytes] = get_collection_values(collection_id)
    return len(collection_values)


@public
def get_collection_values(collection_id: bytes) -> [bytes]:
    """
    Gets the values of a collection
    :param collection_id: The collection_id to get the values of
    :return: An array containing the values of a collection
    """
    collection: Collection = get_collection_internal(collection_id)
    collection_values: [bytes] = collection.get_values()
    return collection_values


@public
def map_bytes_onto_collection(collection_id: bytes, entropy: bytes) -> bytes:
    """
    Returns the value of a collection at an index defined by byte entropy which has been mapped onto the array.
    :param collection_id: The collection_id of the collection to sample from
    :param entropy: The entropy value to sample
    :return: The bytes at the sample location
    """
    collection: Collection = get_collection_internal(collection_id)
    vals: [bytes] = collection.get_values()
    idx: int = Dice.map_bytes_onto_range(0, len(vals) - 1, entropy)
    return vals[idx]


@public
def sample_from_collection(collection_id: bytes, samples: int) -> [bytes]:
    """
    Gets values from a uniform-random sampled index of a colection
    :param collection_id: The collection_id of the collection to sample from
    :param samples: the number of samples to return
    :return: The value at the sampled index
    """
    collection: Collection = get_collection_internal(collection_id)
    collection_values: [bytes] = collection.get_values()

    result: [bytes] = []
    for x in range(samples):
        idx: int = Dice.rand_between(0, len(collection_values) - 1)
        result.append(collection_values[idx])
    return result


@public
def sample_from_runtime_collection(vals: [bytes], samples: int, pick: bool) -> [bytes]:
    """
    Gets values from a uniform-random sampled index an array.  The user has the option of replacement.
    :param vals: the values to uniformly sample from
    :param samples: the number of samples to return
    :return: The value at the sampled index
    """
    assert (not pick) or \
           (len(vals) >= samples)

    result: [bytes] = []
    for x in range(samples):
        idx: int = Dice.rand_between(0, len(vals) - 1)
        result.append(vals[idx])

        if pick:
            # if pick, remove the index as a selectable option
            vals.pop(idx)

    return result


@public
def total_collections() -> int:
    """
    Gets the total collection count
    :return: An integer representing the total collections in the contract
    """
    return total_collections_internal()


@public
def update(script: bytes, manifest: bytes, data: Any):
    """
    Updates the smart contract script
    :param script: The new script to update to
    :param manifest: The new manifest to update to
    :param data: additional data field
    :return:
    :raise AssertionError: raised if the user lacks the "update" permission
    """
    tx = cast(Transaction, script_container)
    signer: UInt160 = tx.sender

    owner: UInt160 = get(OWNER_KEY)

    assert owner == signer, "User Permission Denied"

    update_contract(script, manifest, data)


@public
def _deploy(data: Any, update: bool):
    if not update:
        tx = cast(Transaction, script_container)
        signer: UInt160 = tx.sender
        put(OWNER_KEY, signer)


@contract('0x16d6a0be0506b26e0826dd352724cda0defa7131')
class Dice:

    @staticmethod
    def rand_between(start: int, end: int) -> int:
        pass

    @staticmethod
    def map_bytes_onto_range(start: int, end: int, entropy: bytes) -> int:
        pass

